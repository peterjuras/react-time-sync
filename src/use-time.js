import { useContext, useEffect, useState } from "react";
import TimeContext from "./context";

export function useTime(timerConfig = {}) {
  const timeSync = useContext(TimeContext);
  if (!timeSync) {
    throw new Error(
      "Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"
    );
  }

  const [time, setTime] = useState(() => timeSync.getCurrentTime(timerConfig));

  useEffect(() => timeSync.addTimer(setTime, timerConfig), [
    timerConfig.unit,
    timerConfig.interval
  ]);

  return time;
}
