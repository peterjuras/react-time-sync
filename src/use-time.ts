import { useContext, useDebugValue, useEffect } from "react";
import TimeContext from "./context";
import { Interval } from "time-sync/constants";
import { useStateWithDeps } from "use-state-with-deps";

export interface TimerConfig {
  interval?: Interval;
  unit?: number;
}

export function useTime(timerConfig: TimerConfig = {}): number {
  const timeSync = useContext(TimeContext);
  const [time, setTime] = useStateWithDeps(
    () =>
      timeSync.getCurrentTime({
        interval: timerConfig.interval,
        unit: timerConfig.unit,
      }),
    [timerConfig.interval, timerConfig.unit]
  );

  useEffect(
    (): (() => void) =>
      timeSync.addTimer((currentTime): void => {
        setTime(currentTime);
      }, timerConfig),
    [timerConfig.unit, timerConfig.interval]
  );

  useDebugValue(time);
  return time;
}
