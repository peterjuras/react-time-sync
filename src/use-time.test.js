import { mount } from "enzyme";

import React from "react";
import PropTypes from "prop-types";
import { useTime } from "./use-time";
import TimeSync from "time-sync";
import lolex from "lolex";
import TimeProvider from "./time-provider";

describe("#useTime", () => {
  let clock;

  beforeEach(() => {
    clock = lolex.install({ now: 1 });
  });

  afterEach(() => {
    clock.uninstall();
  });

  it("should be exported correctly", () => expect(useTime).toBeDefined());

  it("should throw if context is not found", () => {
    function TestComponent() {
      const time = useTime();
      return <div>{time}</div>;
    }
    expect(() => {
      const ref = mount(<TestComponent />);
      ref.unmount();
    }).toThrowErrorMatchingSnapshot();
  });

  it("should respect prop updates", () => {
    let renderCalledCount = 0;

    function TestComponent({ unit, interval }) {
      renderCalledCount++;
      const time = useTime({ unit, interval });

      return <div>{time}</div>;
    }
    TestComponent.propTypes = {
      unit: PropTypes.number,
      interval: PropTypes.string
    };
    TestComponent.defaultProps = {
      unit: undefined,
      interval: undefined
    };

    const ref = mount(
      <TimeProvider>
        <TestComponent />
      </TimeProvider>
    );

    expect(ref).toMatchSnapshot();

    clock.tick(999);
    ref.update();
    expect(ref).toMatchSnapshot();

    clock.tick(1000);
    ref.update();
    expect(ref).toMatchSnapshot();

    const newProps = {
      unit: 5,
      interval: TimeSync.MINUTES
    };

    ref.setProps({
      children: React.cloneElement(ref.props().children, newProps)
    });

    ref.update();
    expect(ref).toMatchSnapshot();

    clock.tick(1000 * 5 * 60);

    ref.update();
    expect(ref).toMatchSnapshot();

    ref.unmount();

    expect(renderCalledCount).toBe(5);
  });
});
