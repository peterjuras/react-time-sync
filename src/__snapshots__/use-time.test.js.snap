// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#useTime should respect prop updates 1`] = `
<TimeProvider>
  <TestComponent>
    <div>
      0
    </div>
  </TestComponent>
</TimeProvider>
`;

exports[`#useTime should respect prop updates 2`] = `
<TimeProvider>
  <TestComponent>
    <div>
      1
    </div>
  </TestComponent>
</TimeProvider>
`;

exports[`#useTime should respect prop updates 3`] = `
<TimeProvider>
  <TestComponent>
    <div>
      2
    </div>
  </TestComponent>
</TimeProvider>
`;

exports[`#useTime should respect prop updates 4`] = `
<TimeProvider>
  <TestComponent
    interval="m"
    unit={5}
  >
    <div>
      0
    </div>
  </TestComponent>
</TimeProvider>
`;

exports[`#useTime should respect prop updates 5`] = `
<TimeProvider>
  <TestComponent
    interval="m"
    unit={5}
  >
    <div>
      300
    </div>
  </TestComponent>
</TimeProvider>
`;

exports[`#useTime should throw if context is not found 1`] = `"Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"`;
